tests/cpp/socket_communicator_test.cc

```diff
@@ -73,21 +73,24 @@ TEST(SocketCommunicatorTest, SendAndRecv) {
 #endif  // WIN32
 
 void start_client() {
-  const char * msg = "123456789";
-  sleep(1);
-  SocketSender sender;
-  sender.AddReceiver("127.0.0.1", 2049, 0);
+  sleep(1); // wait server start
+  SocketSender sender(500*1024);
+  sender.AddReceiver("socket://127.0.0.1:50091", 0);
   sender.Connect();
-  sender.Send(msg, 9, 0);
+  for (int i = 0; i < 100; ++i) {
+    sender.Send("123456789", 9, 0);
+  }
   sender.Finalize();
 }
 
 bool start_server() {
-  char serbuff[10];
-  memset(serbuff, '\0', 10);
-  SocketReceiver receiver;
-  receiver.Wait("127.0.0.1", 2049, 1, 500 * 1024);
-  receiver.Recv(serbuff, 9);
+  SocketReceiver receiver(500*1024);
+  receiver.Wait("socket://127.0.0.1:50091", 1);
+  int64_t size = 0;
+  char* data = nullptr;
+  for (int i = 0; i < 100; ++i) {
+    data = receiver.RecvFrom(&size, 0);
+  }
   receiver.Finalize();
-  return string("123456789") == string(serbuff);
+  return string("123456789") == string(data, size);
```

Comment:
please test the code with a more normal scenario. For example, construct NodeFlows, send and receive them and see how it works. It seems the current implementation has memory leak and potential memory error if the sender destroys memory before the message is sent.

Reply:
I will add more test in socket_communicator_test.cc, and I test sending nodeflow in python.